airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+            return false;
+        }
+
+        // Log what we're about to do and determine if we should perform a controlled update
+        if ("1.8.0".equals(currentVersion) || "1.8.1".equals(currentVersion)) {
+            // If we're on a 1.8.0 or 1.8.1 database and upgrading to 2.x, we're going to handle this manually and check what we're doing.
+            LOG.info("HSQLDB database upgrade needed, from version {} to {}.", currentVersion, driverVersion);
+            return true;
+        } else if (currentVersion.startsWith(driverVersion)) {
+            // If we're already on the same version as the driver, nothing should happen.
+            LOG.info("HSQLDB database upgrade unneeded, already on version {}.", driverVersion);
+            return false;
+        } else if (currentVersion.startsWith("2.")) {
+            // If the database version is 2.x but older than the driver, the upgrade should be relatively painless.
+            LOG.info("HSQLDB database will be silently upgraded from version {} to {}.", currentVersion, driverVersion);
+            return false;
+        } else {
+            // If this happens we're on a completely untested version and we don't know what will happen.
+            LOG.warn("HSQLDB database upgrade from version {} to version {} is untested. Trying to connect anyway, this may upgrade the database.", currentVersion, driverVersion);
+            return false;
+        }
+    }
+
+    /**
+     * Perform a backup of the HSQLDB database, to a timestamped directory.
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        Path source = Paths.get(settingsService.getDefaultJDBCPath()).getParent();
+        Path destination = Paths.get(String.format("%s.backup.%s", settingsService.getDefaultJDBCPath(), timestamp));
+
+        LOG.info("Backing up HSQLDB database to {}...", destination.toString());
+        try (Stream<Path> stream = Files.walk(source)) {
+
+            List<Path> sources = stream.collect(toList());
+            List<Path> destinations = sources.stream()
+                    .map(source::relativize)
+                    .map(destination::resolve)
+                    .collect(toList());
+
+            for (int i = 0; i < sources.size(); i++) {
+                LOG.info("Copying {} to {}", sources.get(i), destinations.get(i));
+                Path currentSource = sources.get(i);
+                Path currentDestination = destinations.get(i);
+                if (currentDestination.toFile().isDirectory()) {
+                    currentDestination.toFile().mkdirs();
+                } else {
+                    currentDestination.getParent().toFile().mkdirs();
+                    Files.copy(sources.get(i), destinations.get(i));
+                }
+            }
+        }
```

Comment:
I think you're just trying to copy a folder over here?
We can replace all this by just using `FileUtils.copyDirectory(source,destination)` ?

Reply:
Additionally, out of curiosity, why not use the BACKUP db command?
```
BACKUP DATABASE TO <file path> SCRIPT
```
if the file path is a folder, it'll checkpoint, compact, name and gzip a file in the folder. The name will have the date, and the name of the db, which sounds like what is happening here?

one concern with the manual file copying is the possibility of a file lock, which might be held by something else, possibly the db itself when the file copying is done?

Reply:
I guess I spoke out of context, it seems the shutdown command is issued before a copying is done. Regardless, BACKUP allows the db itself to generate the backup.

Reply:
PS, I have no idea if BACKUP is an option in 1.8. Maybe check first?

Reply:
The big issue about using `BACKUP` at this point is that, when you're connected to the db, it's already too late, the database has been upgraded. And I looked everywhere, but I haven't found a way to make HSQLDB do upgrades other than actually connecting to the database...

Reply:
On the other hand I never knew about `FileUtils.copyDirectory` (why don't these SO answers mention that instead of tens of lines of code if it exists?). Thanks for mentioning!